### PR TITLE
OOP school library: set up associations

### PR DESCRIPTION
**Pull request by yashodhicy**
🐝 Create a class Classroom with the following:

- @label instance variable, should be initialized in the constructor.
- Setter and getter for @label

🐝Create the has-many/belongs-to relationship between Classroom and Student. 

- Create the has-many side (a classroom has many students).
- Create the belongs-to side (a student belongs to a classroom).
- When adding a student to a classroom it also sets the classroom for the student.
- When setting the classroom for a student it also adds it to the classrooms' students.
- @label instance variable initialized in the constructor.
- Setter and getter for @label 

🐝 Create a class Book with the following:

- @title and @author instance variables initialized in the constructor.
- Setters and getters for instance variables 

🐝 Create a class Rental with the following:

- @date instance variable, should be initialized in the constructor.
- Setter and getter for @date .

🐝Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental.

- Create the has-many side of Book and Rental (a book has many rentals).
- Create the belongs-to side of Rental and Book (a rental belongs to a book).
- Create the has-many side of Person and Rental (a person has many rentals).
- Create the belongs-to side of Rental and Person (a rental belongs to a person).
- Modify the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: student.rb [hunk 1]
```diff
@@ -9,6 +9,7 @@ def initialize(age, classroom, **defaults)
 
     super(age, **defaults)
     @classroom = classroom
+    classroom.add_student(self)
   end
```
Comment (mckent05) [inline student.rb]: - Good job done so far, but you have not properly setup the `belongs-to` relationship between the Student class and the Classroom class. According to the project requirements, you should have a custom setter that should look like this:
 
```ruby
def classroom=(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
```

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+  end
```
Comment (mckent05) [inline book.rb]: - Good job done so far, but you have not correctly implemented your add_rental method. Your method should create and new Rental using the following parameters, `book` class, `person` class, and the date. Please fix this 🙏

File: rental.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class Rental
+  attr_accessor :date, :book, :person
+
+  def initialize(date, book, person)
+    @date = date
+    @book = book
+    @person = person
+    book.add_rental(self)
+    person.add_rental(self)
```
Comment (mckent05) [inline rental.rb]: - Good job done so far, but you have not properly implemented the `belong-to` association between the Rental class and the Book and Person class. For example, it should look like this:
`book.rentals << self`. Please fix this for the Person class 👍 👍

File: person.rb [hunk 4]
```diff
@@ -27,6 +28,10 @@ def can_use_services?
   def correct_name
     @name
   end
+
+  def add_rental(rental)
+    @rentals << rental
+  end
```
Comment (mckent05) [inline person.rb]: - Good job done so far, but you have not correctly implemented your add_rental method. Your method should create and new Rental using the following parameters, book class, person class, and date. Please fix this 🙏

File: book.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
```
Comment (mckent05) [inline book.rb]: - According to the project requirements, please implement a getter and setter method for the rental attribute 👍 👍

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @yashodhicy, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
almost there

## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👍 👍 
- Good job setting associations between your classes 👍 👍 


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: student.rb (hunk 1 above)
Comment (yashodhicy) [reply student.rb]: done

File: student.rb [hunk 6]
```diff
@@ -9,6 +9,7 @@ def initialize(age, classroom, **defaults)
 
     super(age, **defaults)
     @classroom = classroom
+    classroom.add_student(self) unless classroom.students.include?(self)
```
Comment (muneebulrehman) [inline student.rb]: As per the project requirements, you need to have a separate method `def classroom=(classroom)` that should serve as the **belongs-to**  side (a student belongs to a classroom). But you have done it in the constructor of the student class. Kindly add a new method as described above for this purpose. Refer to the previous reviewer's example for clarification.

**Review by muneebulrehman — CHANGES_REQUESTED** [review]
Hi @yashodhicy , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
- No linter errors
- Descriptive PR
- Good follow-up with the previous reviewer.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by uwadonat — APPROVED** [review]
## Hi @yashodhicy,

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
### Congratulations! 🎉 
### Highlights
- Detailed pull request message ✔️
- No linter errors ✔️
- Well-organized project and documented project ✔️
- Good commit messages ✔️
- Professional readme file ✔️
# Status: Approved ✔️✔️✔️

209362895-9762f2f8-13cf-4ce0-90b6-a672c4f04196




Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100%.
Please, remember to tag me @uwadonat in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

**Comment by yashodhicy** [comment]
Thank you so much you made my day
@uwadonat

**Pull request merged**